### PR TITLE
WPD-126

### PR DESCRIPTION
**Pull request by d-wysocki**
image

image

https://kindergarten.stxnext.pl/browse/WDP-126

### CODE REVIEW

File: src/presence_analyzer/templates/base.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+     <script src="{{ url_for('static', filename='js/jquery.min.js') }}"></script>
+    <script type="text/javascript" src="https://www.google.com/jsapi"></script>
+
+    {% block head %}  {% endblock %}
+</head>
+
+<body>
+
+    {% block body %}  {% endblock %}
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: remove

File: src/presence_analyzer/templates/base.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    <link href="{{ url_for('static', filename='css/main.css') }}" media="all" rel="stylesheet" type="text/css" />
+
+     <script src="{{ url_for('static', filename='js/jquery.min.js') }}"></script>
+    <script type="text/javascript" src="https://www.google.com/jsapi"></script>
+
+    {% block head %}  {% endblock %}
+</head>
+
+<body>
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: remove

File: src/presence_analyzer/templates/base.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    {% block body %}  {% endblock %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: `{% block body %}{% endblock %}`

File: src/presence_analyzer/templates/base.html [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<!doctype html>
+<html lang=en>
+<head>
+    <meta charset=utf-8>
+    {% block title %} {% endblock %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: remove space between

File: src/presence_analyzer/templates/mean_time_weekday.html [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                <div id="chart_div" style="display: none">
+                </div>
+                <div id="loading">
+                    <img src="{{ url_for('static', filename='img/loading.gif') }}" />
+                </div>
+            </p>
+        </div>
+
+    </div>
+{% endblock %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/mean_time_weekday.html]: This body is repeated in every template.

File: src/presence_analyzer/templates/mean_time_weekday.html [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+{% block body %}
+    <div id="main">
+        <div id="header">
+            <h1>Presence analyzer</h1>
+            <ul>
+                <li><a href="/presence_weekday.html">Presence by weekday</a></li>
+                <li id="selected"><a href="/mean_time_weekday.html">Presence mean time</a></li>
+                <li><a href="/presence_start_end.html">Presence start-end</a></li>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/mean_time_weekday.html]: Do it in a loop in one place.

File: src/presence_analyzer/views.py [hunk 7]
```diff
@@ -19,14 +21,26 @@
...
-    return redirect('/static/presence_weekday.html')
+    return render_template('presence_weekday.html')
+
+
+@app.route('/<template>')
+def dynamic_routes(template):
+    """
+    Create dynamic routes and render template.
+    """
+    return render_template(template)
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: And what if template does not exists?

File: src/presence_analyzer/views.py [hunk 8]
```diff
@@ -48,6 +62,7 @@ def mean_time_weekday_view(user_id):
     """
     Returns mean presence time of given user grouped by weekday.
     """
+    print url_for('mean_time_weekday_view', user_id=user_id)
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: ?

File: src/presence_analyzer/views.py [hunk 9]
```diff
@@ -92,4 +107,5 @@ def presence_start_end(user_id):
     if user_id not in data:
         log.debug('User %s not found!', user_id)
         abort(404)
-    return star_end_time(data, user_id)
+    result = star_end_time(data, user_id)
+    return result
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: What for? Dont assign it if its used only once.

File: src/presence_analyzer/views.py [hunk 10]
```diff
@@ -19,14 +21,26 @@
 )
 
 log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+env = Environment(
+    loader=PackageLoader('presence_analyzer', 'templates'),
+    autoescape=select_autoescape(['html', 'xml'])
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: xml?

You could use a simple dict for it too.

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/views.py [hunk 11]
```diff
@@ -20,13 +20,30 @@
...
+
+
+@app.route('/<template>')
+def dynamic_routes(template):
+    """
+    Create dynamic routes and render template.
+    """
+    try:
+        return render_template(template, pages=links)
+    except Exception:
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: U shouldn't catch generic exception. find out what exception is thrown when there's no such template.

File: src/presence_analyzer/templates/presence_start_end.html [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{% extends "base.html" %}
+{% block head %}
+    {% block title %}<title>Presence analyzer</title>{% endblock %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_start_end.html]: `<title>` shouldn't be wrapped in this block. No need for repeating this in all templates.

File: src/presence_analyzer/templates/presence_start_end.html [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{% extends "base.html" %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_start_end.html:1]: enter between blocks in all templates.

File: src/presence_analyzer/templates/base.html [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+                {%- else -%}
+                <li><a href="{{link}}">{{name}}</a></li>
+                {%- endif -%}
+            {%- endfor -%}
+            </ul>
+        </div>
+        <div id="content">
+            {%- for name, link in pages.items() -%}
+                {%- if link == '/'+request.url.rsplit('/', 1)[1] -%}
+                <h2>{{name}}</h2>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: `{{ name }}`
Indent

Wont be easier to send current pair in context to the template and just apply it here and in generating menu items and adding active class? You're repeating this loop.

File: src/presence_analyzer/templates/base.html [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    {% block head %}{% endblock %}
+</head>
+
+<body>
+<div id="main">
+        <div id="header">
+            <h1>Presence analyzer</h1>
+            <ul>
+            {%- for name, link in pages.items() -%}
+                {%- if link == '/'+request.url.rsplit('/', 1)[1] -%}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: What for you need to add this leading slash?
Do it without additional  param in rsplit, it's not needed, and select the last element, no the second one.
And that if should only add the class, not generate the whole `<li>`.

File: src/presence_analyzer/templates/base.html [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+                <li id="selected"><a href="{{link}}">{{name}}</a></li>
+                {%- else -%}
+                <li><a href="{{link}}">{{name}}</a></li>
+                {%- endif -%}
+            {%- endfor -%}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: indent

File: src/presence_analyzer/templates/base.html [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+<body>
+<div id="main">
+        <div id="header">
+            <h1>Presence analyzer</h1>
+            <ul>
+            {%- for name, link in pages.items() -%}
+                {%- if link == '/'+request.url.rsplit('/', 1)[1] -%}
+                <li id="selected"><a href="{{link}}">{{name}}</a></li>
+                {%- else -%}
+                <li><a href="{{link}}">{{name}}</a></li>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: indent

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/templates/presence_weekday.html [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{% extends "base.html" %}
+
+    {% block title %}Presence analyzer{% endblock %}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_weekday.html]: I don't see the difference between those.

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/templates/base.html [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        <script type="text/javascript" src="https://www.google.com/jsapi"></script>
+        {% block head %}{% endblock %}
+    </head>
+    <body>
+        <div id="main">
+            <div id="header">
+                <h1>Presence analyzer</h1>
+                <ul>
+                {%- for link, name in pages.items() -%}
+                    {%- if link == request.url.rsplit('/')[-1] -%}
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: Why not simple endswith?

File: src/presence_analyzer/templates/base.html [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+                        <li id="selected">
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: Only `id="selected"` differs.

File: src/presence_analyzer/templates/base.html [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+                    {%- else -%}
+                        <li>
+                    {%- endif -%}
+                        <a href="{{link}}">{{name}}</a></li>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/base.html]: There was a comment already about it, READ IT!

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: buildout.cfg [hunk 22]
```diff
@@ -90,6 +90,7 @@ defaults = -v
 
 [pep8]
 recipe = zc.recipe.egg
+
```
Comment (dpolcyn) [inline buildout.cfg]: ???

File: mean_time.js [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+                    $loading.show();
+                    $chartDiv.hide();
+
+                    $.getJSON("/api/v1/mean_time_weekday/" + $selectedUser, function(result) {
+                        var data = new google.visualization.DataTable(),
+                            options = {
+                                hAxis: {title: 'Weekday'}
+                            },
+                            formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'}),
+                            chart = new google.visualization.ColumnChart($chartDiv[0]);
```
Comment (dpolcyn) [inline mean_time.js]: multiple variable declaration :+1:

File: src/presence_analyzer/templates/base.html [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+                <div id="loading">
+                    <img src="{{ url_for('static', filename='img/loading.gif') }}" />
+                </div>
+                </p>
+            </div>
+        </div>
+    </body>
+</html>
+
+
```
Comment (dpolcyn) [inline src/presence_analyzer/templates/base.html]: To many empty lines

File: src/presence_analyzer/tests.py [hunk 25]
```diff
@@ -152,6 +150,18 @@ def test_presence_start_end(self):
             }
         )
 
+    def test_dynamic_route(self):
+        """
+        Testing creating dynamic routes and templates
+        """
+        resp = self.client.get('/mean_time_weekday.html')
+
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Without empty line here

File: src/presence_analyzer/tests.py [hunk 26]
```diff
@@ -152,6 +150,18 @@ def test_presence_start_end(self):
...
+        self.assertEqual(resp.status_code, 200)
+
+        bad_resp = self.client.get('/something.html')
+
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 27]
```diff
@@ -266,7 +271,7 @@ def test_average_second(self):
         self.assertEqual(
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py:272]: Style

**Pull request merged**